### ISSUE TITLE
[V2] :global doesn't rerender. :global stay on the page when component is unmounted

### ISSUE THREAD

**Issue by kopax** [styled-components/styled-components#612]
I have found a bug in `2.0.0-9` using global injection.

When I use **:global** within a component for example :

```jsx
styled.div`
    :global(body) {
      ${props.isOpen ? `margin-right: ${props.size}px !important;` : ''}
    }

    ${!props.isOpen ? `transform: translate3d(${props.size-10}px, 0, 0);` : ''}
`
```
## Version
2.0.0-8

## Reproduction

Unfortunately it's hard to produce a reproduction, webpackbin work bad these days. 
I don't know how to use styled template literal in codepen.

## Steps to reproduce

Using **react-router**, create a router that container a css injection that use **:global**. 
Go on that page, then change to another page, the css stay injected

## Expected Behavior

- When the component is rerendered with a new `props.size` **:global** should apply the new css
- if the components get umounted, the style injected in body are removed from on the page.

## Actual Behavior

- When the component is rerendered with a new `props.size` **:global** doesn't refresh the css
- if the components get umounted, the style injected in body stay on the page.

**Comment by geelen** [comment]
Um... what is `:global`? That's not a thing in Styled Components (unless someone merged something without me noticing...)

**Comment by k15a** [comment]
@geelen that's part of stylis, the new parser Max merged into version 2.

**Comment by kopax** [comment]
It also appear to not be removed when using `@global` in plus of `:global()`.

Fwd @thysultan, is it stylis or styled issue ?

**Comment by thysultan** [comment]
If it's rendered the first time i don't think it is a stylis issue.

**Comment by kopax** [comment]
Which means there are probably more `v2` features that might not be removed (@mixin, @include, @global, :global, :host, :host(), :host-context()).

I am not used with type screen and the code flow of styled. Where does the remove part happen ?

**Comment by philpl** [comment]
@kopax styled-components doesn't remove any CSS rules, once they were injected. Thus this will likely never work in this way

**Comment by kopax** [comment]
@philpl thanks for the details. If styled-components doesn't remove any CSS rules once injected, it refresh the CSS.

This way we can update the theme dynamically using a `<ThemeProvider />`

The problem is that the css doesn't get re-rendered and stay in the page. 

This corrupt the GUI when the ThemeProvider change the theme.

**Comment by geelen** [comment]
> Which means there are probably more v2 features that might not be rerendered and stays in the page (@mixin, @include, @global, :global, :host, :host(), :host-context()).

@mxstbr can you weigh in here? I thought we weren't adding any of these?

For the record, I'm extremely against _any_ CSS metaprogramming in Styled Components without _very_ careful consideration. At the moment, the only two things we allow are nesting (`&`) and autoprefixing. So just because Stylis or PostCSS supports something, doesn't mean we will expose it. You have all of JS to build mixins, and `injectGlobal` for global rules, anything that diverges from our simplified superset of raw CSS makes it vastly more difficult to build tooling like @philpl's preprocessor in the future.

**Comment by philpl** [comment]
@geelen Specifically some of stylis' features should be switched off, some not:

```stylis(
    selector: {string},     // selector - i.e `.class` or `#id` or `[attr=id]`
    styles: {string},       // css string
    animations: {boolean=}  // false to prevent prefixing animations, true by default
    compact: {boolean=}     // enable additional features (mixins, variables)
    middleware: {function=}
);
```

We're switching off the animations and compact part. (Although `compact` seems confusing. Does it really switch `mixins` and `variables` *off*?)

There's no mention in the docs whether we can switch off `:global`, `@global` and `:host` though.

cc @thysultan

**Comment by thysultan** [comment]
Yes it does disable mixins and variables, planning on a release that removes them, since you could use middleware to re-construct them, `:global` selector is not, will improve the compact option to include it with that release.

**Comment by geelen** [comment]
Yeah if we can remove `:global`, `@global` and `:host` and let people use `&` and `injectGlobal` that would be great, thanks @thysultan!

**Comment by kopax** [comment]
@geelen @thysultan @philpl I understand that every over variable should be disabled.

I have just one request is to keep the `:global` available. This allow us sibling for parent component and many times it is required to add a class to the `<body>` or to a wrapper. This logic should be wrapped in the component itself not the application.

It makes a huge difference to remove it. Also, `injectGlobal` is here because `:global` wasn't existing before.

Otherwise, would it be possible to create a special component that can take care of that ? (<ThemeProvider />) for example ? In case we still wan't to rerender the global with a new theme ?

**Comment by geelen** [comment]
You can do global selectors like things on the `body` just fine:

```js
styled.div`
  body.xyz & {
    /* ... */
  }
`
```

No `:global` needed. The only restriction is that the rule _must_ include a `&`—i.e. every rule for a component _must_ be related to that component in some way. That's the restriction I don't want to relax.

**Comment by kopax** [comment]
How about if I need to rerender the body part (font) using the ThemeProvider ? is this possible ?

I guess it would be possible with the same technique you just give me but it needs to be inside this wrapper right ?

**Comment by geelen** [comment]
Not sure what you mean by the "wrapper". Not sure what you might want controlled by a theme that's _also_ global. If you had some `Root` element, you don't need any global selectors at all:

```js
const Root = styled.main`
  font-family: ${ props => props.theme.fontFamily };
`
```

Normal CSS inheritance will see that flow down to every component nested inside it.

**Comment by cpsubrian** [comment]
Not sure if this is the best place for this comment, but just want to weigh in that (as far as I can tell), `injectGlobal` does not work with the `<ThemeProvider>` system or auto-prefixing at all. Since I often convert `html,body` to be display flex (and perhaps a theme would want to set the bg color), I'd love to hear about solutions for that.

**Comment by mxstbr** [comment]
I agree 100% we definitely don't want any of that extra stuff, I didn't realise that slipped through, sorry! Is that release out yet @thysultan?

**Comment by mxstbr** [comment]
@cpsubrian yes `injectGlobal` doesn't get the theme, but it should be auto-prefixed. If it isn't that's a bug!

**Comment by thysultan** [comment]
@mxstbr included in v2.0.0
